compiler/src/main.cc

```diff
@@ -898,40 +908,49 @@ bool skip_utf8_bom(FILE* f) {
 void parse(t_program* program, t_program* parent_program) {
   // Get scope file path
   string path = program->get_path();
+  string parent_path = (parent_program != NULL) ? parent_program->get_path() : "null";
 
-  // Set current dir global, which is used in the include_file function
-  g_curdir = directory_name(path);
-  g_curpath = path;
-
-  // Open the file
-  // skip UTF-8 BOM if there is one
-  yyin = fopen(path.c_str(), "r");
-  if (yyin == 0) {
-    failure("Could not open input file: \"%s\"", path.c_str());
-  }
-  if (skip_utf8_bom(yyin))
-    pverbose("Skipped UTF-8 BOM at %s\n", path.c_str());
+  std::map<std::string, t_scope>::iterator it = program_to_scope.find(path);
+  if (it != program_to_scope.end()) {
+    pverbose("Skipping recursion for parsed program %s\n", path.c_str());
+    *(program->scope()) = it->second;
+  } else {
+    // Set current dir global, which is used in the include_file function
+    g_curdir = directory_name(path);
+    g_curpath = path;
+
+    // Open the file
+    // skip UTF-8 BOM if there is one
+    yyin = fopen(path.c_str(), "r");
+    if (yyin == 0) {
+      failure("Could not open input file: \"%s\"", path.c_str());
+    }
+    if (skip_utf8_bom(yyin))
+      pverbose("Skipped UTF-8 BOM at %s\n", path.c_str());
+
+    // Create new scope and scan for includes
+    pverbose("Scanning includes from %s \n", path.c_str());
+    g_parse_mode = INCLUDES;
+    g_program = program;
+    g_scope = program->scope();
+    try {
+      yylineno = 1;
+      if (yyparse() != 0) {
+        failure("Parser error during include pass.");
+      }
+    } catch (string x) {
+      failure(x.c_str());
+    }
+    fclose(yyin);
 
-  // Create new scope and scan for includes
-  pverbose("Scanning %s for includes\n", path.c_str());
-  g_parse_mode = INCLUDES;
-  g_program = program;
-  g_scope = program->scope();
-  try {
-    yylineno = 1;
-    if (yyparse() != 0) {
-      failure("Parser error during include pass.");
+    // Recursively parse all the include programs
+    vector<t_program*>& includes = program->get_includes();
+    vector<t_program*>::iterator iter;
+    for (iter = includes.begin(); iter != includes.end(); ++iter) {
+      parse(*iter, program);
     }
-  } catch (string x) {
-    failure(x.c_str());
-  }
-  fclose(yyin);
 
-  // Recursively parse all the include programs
-  vector<t_program*>& includes = program->get_includes();
-  vector<t_program*>::iterator iter;
-  for (iter = includes.begin(); iter != includes.end(); ++iter) {
-    parse(*iter, program);
+    program_to_scope[path] = *(program->scope());
```

Comment:
this only happens if we _haven't yet_ found the path, at which point we're (a) indicating that we just parsed this, and (b) storing the current scope